### DISCUSSION TITLE
Q: How To Execute Javascript In A Chrome Tab

### DISCUSSION THREAD

**Discussion by muescha** [Hammerspoon/hammerspoon#3380] [Q&A]
I like to know some ideas and some hints in which direction I can dig to find a solution for this

## given:
- I have a Hammerspoon script to control video stop/pause - I do this sending keystrokes to an app - for example `space` to pause a video
- Sometimes it not only stop the video - instead it also execute the scroll the page which is also assigned to `space` in chrome

## what i like to do:
- I like to find a way to select the player div in a website so that my kontrol keys are executed in scope of the (then) selected player

## bonus:
- i can direct call some javascript functions of some injected players which are assigned to buttons and player controls

**Comment by megalithic** [comment]
Not sure if this fully solves it, but you can, with JXA, select a tab and do something in that tab with JS, so basically at that point, you're in the `window` context of that page:

https://github.com/megalithic/dotfiles/blob/main/config/hammerspoon/lib/browser.lua#L40-L57

**Reply by muescha** [reply]
thanks for the idea - but it seems that with JXA it is only possible to control some apps but I can not go deep into the window context?

**Reply by muescha** [reply]
but maybe you are right - it can be possible if i see this code snippet:
```javascript
const tab = Application('Google Chrome').windows[0].activeTab()
tab.execute({
  javascript: 'document.querySelector("#dns-view-clear-cache").click()'
})
```
https://github.com/JXA-Cookbook/JXA-Cookbook/wiki/Safari-and-Chrome#clear-chromes-dns-cache

```
execute v : Execute a piece of javascript.
   execute specifier : The tab to execute the command in.
      javascript text : The javascript code to execute.
          → any
```

sources:
- https://www.macstories.net/tutorials/getting-started-with-javascript-for-automation-on-yosemite/
- https://hackmag.com/coding/getting-to-grips-with-javascript-automation-for-os-x/
- https://stackoverflow.com/questions/55949657/getting-values-from-inner-functions-of-current-webpage-in-chrome-via-jxa-appl
